### PR TITLE
[CUDA] Add binary objective for cuda_exp

### PR DESCRIPTION
**Pull request by shiyu1994**
Add binary objective for `cuda_exp`. (#5163)

### CODE REVIEW

**Comment by guolinke** [comment]
it seems the cuda self-hosted machine in queue? BTW, do we have a test for the cuda binary objective?

**Comment by shiyu1994** [comment]
> BTW, do we have a test for the cuda binary objective?

I think this can be covered by existing tests for binary objective on CPUs, for example,
https://github.com/microsoft/LightGBM/blob/9489f878b3568e70b441e5df602483e116f24cc6/tests/python_package_test/test_engine.py#L61-L83

because we are replacing the default device with `cuda_exp` in
https://github.com/microsoft/LightGBM/blob/9489f878b3568e70b441e5df602483e116f24cc6/.ci/test.sh#L204
when testing for `cuda_exp` version.

**Comment by shiyu1994** [comment]
It is weird that this PR should now passes all the tests. But my local tests are failing with `cuda_exp` enabled. I'm taking a deeper check into this.

File: src/objective/cuda/cuda_binary_objective.cu [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+  const double sigmoid, const data_size_t num_data,
+  score_t* cuda_out_gradients, score_t* cuda_out_hessians) {
+  const data_size_t data_index = static_cast<data_size_t>(blockDim.x * blockIdx.x + threadIdx.x);
+  if (data_index < num_data) {
+    const label_t cuda_label = static_cast<int>(cuda_labels[data_index]);
+    const int label = IS_OVA ? (cuda_label == ova_class_id ? 1 : -1) : (cuda_label > 0 ? 1 : -1);
+    const double response = -label * sigmoid / (1.0f + exp(label * sigmoid * cuda_scores[data_index]));
+    const double abs_response = fabs(response);
+    if (!USE_WEIGHT) {
+      if (USE_LABEL_WEIGHT) {
```
Comment (guolinke) [inline src/objective/cuda/cuda_binary_objective.cu:129]: should we use `constexpr` for these template conditions ?

Comment (shiyu1994) [reply src/objective/cuda/cuda_binary_objective.cu:129]: constexpr if is only available with c++17 or higher version.

**Pull request merged**